src/untyped_array.rs

```diff
@@ -0,0 +1,254 @@
+//! Safe, untyped interface for NumPy's [N-dimensional arrays][ndarray]
+//!
+//! [ndarray]: https://numpy.org/doc/stable/reference/arrays.ndarray.html
+use std::{os::raw::c_int, slice};
+
+use pyo3::{
+    ffi, pyobject_native_type_extract, pyobject_native_type_named, AsPyPointer, IntoPy, PyAny,
+    PyNativeType, PyObject, PyTypeInfo, Python,
+};
+
+use crate::cold;
+use crate::dtype::PyArrayDescr;
+use crate::npyffi;
+
+/// A safe, untyped wrapper for NumPy's [`ndarray`][ndarray] class.
+///
+/// Unlike [`PyArray<T,D>`][crate::PyArray], this type does not constrain either element type `T` nor the dimensionality `D`.
```

Comment:
I think it's helpful to mention that
- `PyUntypedArray` can be upcasted to `PyArray<T, D>`
- Instances of `PyArray<T, D>` can access `PyUntypedArray`'s methods by Deref